README.md
# SBTS Cloudfiles Uploader
Add wordpress uploads to cloud files, manage links to serve only cdn links to cloud files, and add custom upload manager.

## Installation
In your plugins directory (/wordpress/wp-content/plugins/):  
    $ git clone git@github.com:sbtsdev/sbts-cf-uploader.git  
    $ cd sbts-cf-uploader  
    $ composer install  

## Configuration
1. Log into Wordpress
2. Go to Plugins
3. Enable Cloud Files Uploader
4. Go to Settings > Media
5. Under Cloud Files Settings add your Cloud Files username and API key  
    A. Add your Cloud Files username  
    B. Add your Cloud Files API key  
    C. Save the Media settings  
    D. Add the Cloud Files container that Wordpress will use to upload files to  

## Usage
Wordpress media uploads can be used normally. Uploaded media will be put on Cloud Files and references to Cloud files will
be saved with the media.

Under Media there is also a Cloud Files Uploader. This gives the user the ability to manually upload files as well as view uploaded contents for each container in the Cloud Files account.

## Access
Only users who can upload files are able to use the uploader. This is a default wordpress permission.


CHANGELOG.md
SBTS CF Uploader Version History
================================

Date        Version Explanation
----        ------- -----------
20160418    0.3.0   Updated OpenCloud API; improved js interactions; accounted for WP image choice optimization  
20130809    0.2.2   Allow faster return of containers through caching  
20130801    0.2.1   Added remove_filter for two Root Relative filters that caused Insert into Post to return wrong  
20130723    0.2.0   Migration to php-opencloud API  
                    Added Wordpress upload/sideload option  
                    Added jpg to permitted file types  
20130506    0.1.5   Added custom functionality to store the duration of audio/mpeg files in CF metadata  
                    Added single get_file function for other plugins to use  
                    Improved function structure  
20130502    0.1.4   Allowed uploads to execute as long as needed  
                    Fix for plugin_dir_url php notice  
                    Cosmetic language change, from images to files on template  
20130430    0.1.3   Fixed upload after quick fix for file types (0.1.2 broke it and copy/paste)  
                    Fixed copy and paste via hack to ZeroClipboard instance on mouseenter  
                    Improved download links with download attribute in anchor tag  
                    Fixed fallback for pre-drag-and-drop (or FileReader) browsers  
                    Fixed delete so the file listing was removed upon successful delete  
20130429    0.1.2   Temp fix for uploads so Firefox could do upload; fix to ZeroClipboard to reposition Copy button  
20130318    0.1.1   Fixed bug: make sure we anticipate another plugin using Cloud Files php api  
20130315    0.1.0   Initial use: upload (single and multiple), list, delete, switch containers all working  


settings.php
<?php
/*
 * This file allows the Settings>Media Cloud Files Uploader Settings to function
 */
if (! class_exists( 'SBTS_CF_Plugin_Settings' ) ) {
    class SBTS_CF_Plugin_Settings {

        public function __construct() {
            if (! current_user_can( 'upload_files' ) ) {
                return false;
            }
            // register actions
            add_action( 'admin_init', array( &$this, 'admin_init' ) );
        }

        public function admin_init() {
            // register your plugin's settings
            register_setting( 'media', 'sbts_cf_uploader_username' );
            register_setting( 'media', 'sbts_cf_uploader_api_key' );
            register_setting( 'media', 'sbts_cf_uploader_wp_upload_container' );

            // add your settings section
            add_settings_section(
                'sbts_cf_plugin_settings-section',
                'Cloud Files Settings',
                array( &$this, 'settings_section_sbts_cf_uploader' ),
                'media'
            );

            // add your setting's fields
            add_settings_field(
                'sbts_cf_plugin_settings-sbts_cf_uploader_username',
                'Cloud Files username',
                array( &$this, 'settings_field_input_text' ),
                'media',
                'sbts_cf_plugin_settings-section',
                array(
                    'field' => 'sbts_cf_uploader_username',
                    'type'  => 'text'
                )
            );
            add_settings_field(
                'sbts_cf_plugin_settings-sbts_cf_uploader_api_key',
                'Cloud Files API Key',
                array( &$this, 'settings_field_input_text' ),
                'media',
                'sbts_cf_plugin_settings-section',
                array(
                    'field' => 'sbts_cf_uploader_api_key',
                    'type'  => 'password'
                )
            );
            add_settings_field(
                'sbts_cf_plugin_settings-sbts_cf_uploader_wp_upload_container',
                'Cloud Files container for native Wordpress uploads (sync only server to container not container back to server)',
                array( &$this, 'settings_field_select_container' ),
                'media',
                'sbts_cf_plugin_settings-section',
                array(
                    'field' => 'sbts_cf_uploader_wp_upload_container',
                    'type'  => 'select'
                )
            );
        }

        public function settings_section_sbts_cf_uploader() {
            echo 'Complete these settings to connect Cloud File Uploader to Rackspace Cloud Files';
        }

        /**
         * This function provides text inputs for settings fields
         */
        public function settings_field_input_text( $args ) {
            // Get the field name from the $args array
            $field  = $args['field'];
            $type   = $args['type'];
            // Get the value of this setting
            $value = get_option( $field );
            // echo a proper input type="text"
            echo sprintf( '<input type="%s" name="%s" id="%s" value="%s" />', $type, $field, $field, $value );
        }

        public function settings_field_select_container( $args ) {
            $containers = array();
            if ( class_exists( 'Cloud_Files_Manager' ) ) {
                $cfm = new Cloud_Files_Manager( $this->get_settings() );
                try {
                    $containers = $cfm->get_containers();
                } catch ( Exception $e ) {
                    error_log( 'SBTS CF Uploader: Settings could not retrieve containers for WP upload option. ' . $e );
                }
            }
            if ( count( $containers ) > 0 ) {
                $field  = $args['field'];
                $type   = $args['type'];
                $value  = get_option( 'sbts_cf_uploader_wp_upload_container' );
                echo    sprintf( '<select id="%s" name="%s">', $field, $field );
                echo    sprintf( '  <option value="">Do not upload/sideload Wordpress uploads</option>' );
                foreach ( $containers as $cont ) {
                    $selected = ( $value === $cont['name'] ) ? ' selected="selected"' : '';
                    $size = ( $cont['bytes'] > 1024 * 1024 * 1024 ? round( $cont['bytes'] / ( 1024 * 1024 * 1024 ), 2 ) . ' Gb' : round( $cont['bytes'] / ( 1024 * 1024 ), 2 ) . ' Mb' );
                    echo    sprintf( '  <option value="%s"%s>%s (%s) [%s]</option>', $cont['name'], $selected, $cont['name'], $cont['count'], $size );
                }
                echo    sprintf( '</select>' );
            }
        }

        public function get_settings() {
            return array(
                'username'  => get_option( 'sbts_cf_uploader_username' ),
                'api_key'   => get_option( 'sbts_cf_uploader_api_key' ),
                'wp_upload_container'   => get_option( 'sbts_cf_uploader_wp_upload_container' )
            );
        }
    }
}


cloud-files-manager.php
<?php
require 'vendor/autoload.php';
use \OpenCloud\Rackspace;
if (! class_exists( 'Cloud_Files_Manager' ) ) {
    class Cloud_Files_Manager {
        private $cf_settings;
        private $cf_obj_store_service;
        private $cf_containers;

        function __construct( $settings ) {
            $this->cf_settings = $settings;
            $this->cf_containers = array();
            if (! class_exists( 'Mp3Info' ) ) {
                require_once( 'inc/mp3info.class.php' );
            }
            // Initialize Cloud Files API for Manager to use
            if (! class_exists( 'ClassLoader' ) ) { // play nice with 'other' plugins (and hope they're up to date)
                define( 'RAXSDK_TIMEOUT', 0 ); // we'll wait forever
            }

        }

        public function connect( $force_new = false ) {
            if ( (! $force_new ) && isset( $this->cf_obj_store_service ) ) {
                return $this->cf_obj_store_service;
            }
            $credentials = array(
                'username'  => $this->cf_settings['username'],
                'apiKey'    => $this->cf_settings['api_key']
            );
            try {
                $cf_conn = new Rackspace( Rackspace::US_IDENTITY_ENDPOINT, $credentials );
                $this->cf_obj_store_service = $cf_conn->objectStoreService( null, 'DFW' );
            } catch ( Exception $e ) {
                throw $e;
            }
        }

        public function get_container( $req_container ) {
            $this->connect();
            $container = $this->cf_obj_store_service->getContainer( $req_container );
            return $container;
        }

        public function get_containers() {
            if ( isset( $this->cf_containers ) && (! empty( $this->cf_containers ) ) ) {
                return $this->cf_containers;
            }
            $this->connect();
            $list = $this->cf_obj_store_service->listContainers();
            foreach ( $list as $container ) {
                if ($container->isCdnEnabled()) {
                    $cdn_container = $container->getCdn();
                    try {
                        $cdn_url = $cdn_container->getCdnUri();
                    } catch ( Exceptions $e ) {
                        $cdn_url = '';
                        continue;
                    }
                    $count = $container->getObjectCount();
                    $bytes = $container->getBytesUsed();
                    $this->cf_containers[] = array(
                        'name'  => $cdn_container->name,
                        'count' => $count,
                        'bytes' => $bytes,
                        'url'   => $cdn_url
                    );
                }
            }
            return $this->cf_containers;
        }

        public function get_file( $req_container, $req_file_name ) {
            $file = array();
            try {
                $this->connect();
                $container = $this->cf_obj_store_service->getContainer( $req_container );
                $object = $container->getObject( $req_file_name );
                $file[] = $this->make_file( $object );
            } catch ( Exception $e ) {
                throw $e;
            }
            return $file;
        }

        public function get_files( $req_container, $req_prefix = NULL ) {
            $files = array();
            try {
                $this->connect();
                $container = $this->cf_obj_store_service->getContainer( $req_container );
                $objects = $container->objectList( array( 'prefix', $req_prefix ) );
                foreach ($objects as $object) {
                    // TODO manage the pseudo-directories too
                    if ( 'application/directory' !== $object->getContentType() ) {
                        $files[] = $this->make_file( $object );
                    }
                }
            } catch ( Exception $e ) {
                throw $e;
            }
            return $files;
        }

        public function upload_files( $req_container, $files ) {
            $names = array();
            if ( class_exists( 'Mp3Info' ) ) {
                $mp3_info = new Mp3Info();
            }
            if ( is_array( $files ) ) {
                $this->connect();
                $container = $this->cf_obj_store_service->getContainer( $req_container );
                foreach ( $files as $file ) {
                    try {
                        // should we bother using move_uploaded_file to verify that the file "was uploaded using PHP's HTTP POST upload mechanism"?
                        // try to add audio duration to metadata
                        $metadata = array();
                        if ( $mp3_info && ( substr( $file['name'], -4 ) === '.mp3' ) ) {
                            $mp3_data = $mp3_info->GetMp3Info( $file['location'] );
                            if ( $mp3_data && isset( $mp3_data['playtime_string'] ) ) {
                                $metadata['Rduration'] = $mp3_data['playtime_string'];
                            }
                        }
                        $container->uploadObject( $file['name'], fopen( $file['location'], 'r+' ), $metadata );
                        $names[] = $file['name'];
                    } catch ( Exception $e ) {
                        throw $e;
                    }
                }
                return $this->verify_uploads( $container, $names );
            }
            return $names;
        }

        public function delete_files( $req_container, $files ) {
            $names = array();
            $to_delete = array();
            $deleted = array();
            if ( is_array( $files ) ) {
                foreach ( $files as $file ) {
                    $names[] = $file['name'];
                }
                $this->connect();
                $container = $this->cf_obj_store_service->getContainer( $req_container );
                $to_delete = $this->verify_uploads( $container, $names );
                foreach ( $to_delete as $delete_obj ) {
                    try {
                        $container->deleteObject($delete_obj['full_name']);
                        $deleted[] = $delete_obj;
                    } catch ( Exception $e ) {
                        throw $e;
                    }
                }
            }
            return $deleted;
        }

        private function verify_uploads( $container, $names ) {
            $files = array();
            foreach ( $names as $name ) {
                if ( '/' === substr( $name, 0, 1 ) ) {
                    $name = substr( $name, 1 );
                }
                $obj = $container->getObject( $name );
                $files[] = $this->make_file( $obj );
            }
            return $files;
        }

        private function make_file( $obj ) {
            $fullName = $obj->getName();
            $bg = strrpos( $fullName, '/' );
            $path = '/' . ( $bg ? substr( $fullName, 0, $bg + 1 ) : '' );
            $name = substr( $fullName, $bg ? $bg + 1 : 0 );
            $bytes = $obj->getContentLength();
            $size = round( $bytes / ( 1024.00 * 1024.00 ), 2 );

            $mdata = $obj->getMetadata()->toArray();

            $dt = strtotime( $obj->getLastModified() );
            $last_mod = ( $dt ? date( 'Y-m-d H:i:s', $dt ) : '' );

            $cdn_uri = $obj->getContainer()->getCdn()->getCdnUri();

            return array(
                'full_name' => $fullName,
                'path'      => $path,
                'name'      => $name,
                'uri'       => $cdn_uri. '/' . $fullName,
                'type'      => $obj->getContentType(),
                'rsize'     => $bytes,
                'size'      => $size . " Mb",
                'mdata'     => $mdata,
                'last_mod'  => $last_mod
            );
        }
    }
}
?>


js/sbts-cf-uploader-core-v0.3.js
(function (win, doc, $, undefined) {
    var holder = doc.getElementById('holder'),
    tests = {
        filereader: typeof FileReader != 'undefined',
        dnd: 'draggable' in doc.createElement('span'),
        formdata: !!win.FormData,
        progress: "upload" in new XMLHttpRequest()
    },
    // TODO load valid types via ajax call to wordpress
    accepted_type = {
        '.png'  : true,
        '.jpg'  : true,
        '.jpeg' : true,
        '.gif'  : true,
        '.mpeg' : true,
        '.mp3'  : true,
        '.flv'  : true,
        '.m4a'  : true,
        '.mp4'  : true
    },
    tmpl_jq = '#file_up_tmpl',
    tmpl_st_jq = '#tree_tmpl',
    template = {
        'file_up'   : Handlebars.compile($(tmpl_jq).html()),
        'sub_tree'  : Handlebars.compile($(tmpl_st_jq).html())
    },
    containers = [],
    files = [],
    tree_delimeter = '/', // TODO pull tree_delimeter from server settings
    tree = {},
    tree_root_id = 'sbts_cf_tree',
    tree_root_jq = '#' + tree_root_id,
    path_id_pre = 'sbts_cf_path_id_',
    path_select_id_pre = 'path_select_',
    path_display_jq = '#path_display',
    full_path_jq = '#full_path_path',
    user_added_path_jq = '#additional_path',
    error_no_path = false,
    sub_tree_jq = '.sub-tree',
    cont_pick_jq = '#sbts_cf_uploader_container',
    file_list_jq = '.file-list',
    message_jq = '.activity-response',
    error_class = 'error-text',
    progress_bar_jq = '#ajax_progress>span',
    zc_clip,
    plugin_url = win.sbts_cf_uploader ? win.sbts_cf_uploader.url : ''; // Not a good use of a global, to be fixed when we make this a Backbone script

    function progress_handler (e){
        var perc = e.loaded / e.total * 100;
        if(e.lengthComputable){
            $(progress_bar_jq).css('width', perc + '%');
        }
    }

    // use temp_tree recursively to add tree[tree_delimeter].media.audio.Podcast = {'__files':[4,10,18]}
    // pa should be in the form pa = ['media', 'audio', 'Podcast']
    // every listing should have a path and file
    function add_path_to_tree (temp_tree, pa, i) {
        var pa_current = pa.shift();
        if (! temp_tree[pa_current]) {
            if (pa.length > 0) {
                temp_tree[pa_current] = {};
                temp_tree[pa_current] = add_path_to_tree(temp_tree[pa_current], pa, i);
            } else {
                temp_tree[pa_current] = {'__files':[i]};
            }
        } else {
            if (pa.length > 0) {
                temp_tree[pa_current] = add_path_to_tree(temp_tree[pa_current], pa, i);
            } else {
                if (! temp_tree[pa_current].__files) {
                    temp_tree[pa_current].__files = [i];
                } else {
                    temp_tree[pa_current].__files.push(i);
                }
            }
        }
        return temp_tree;
    }

    function add_files_to_tree (more_files) {
        var i = 0, file_len = more_files.length,
        path, pa, path_len;
        for (; i < file_len; i += 1) {
            path = more_files[i].path || '';
            pa = path.substr(1, path.length - 2).split(tree_delimeter);
            tree = add_path_to_tree(tree, pa, i);
        }
    }

    function load_tree () {
        tree = {};
        tree[tree_delimeter] = {};
        add_files_to_tree(files);
    }

    function get_sub_tree_by_path (path) {
        var i = 0, pa, pa_len, sub_tree = tree;
        if (path !== tree_delimeter) {
            pa = path.substr(1, path.length - 2).split(tree_delimeter);
            pa_len = pa.length;
            for (; i < pa_len; i += 1) {
                sub_tree = sub_tree[pa[i]];
            }
        }
        return sub_tree;
    }

    function display_sub_tree (sub_tree, $target, building_path) {
        $.each(sub_tree, function (index, val) {
            var st_obj, // object to pass to the template
            id_post, // what makes the id's unique
            my_path, // keep track of the path structure for this iteraction
            i = 0;
            // display __files, if any
            if ('__files' === index) {
                files_len = val.length;
                for (i = 0; i < files_len; i += 1) {
                    $target.append(template.file_up(files[val[i]]));
                }
            } else {
                my_path = building_path + (tree_delimeter === index ? index : tree_delimeter === building_path ? index : tree_delimeter + index);
                id_post = my_path.replace(new RegExp(tree_delimeter, 'g'), '_');
                st_obj = {
                    'path_id'   : path_id_pre + id_post,
                    'sel_id'    : path_select_id_pre + id_post,
                    'path'      : my_path
                };
                $target.append(template.sub_tree(st_obj));
                display_sub_tree(val, $target.children('#' + st_obj.path_id).children(sub_tree_jq).first(), my_path);
            }
        });
    }

    function display_tree_by_path (path, target) {
        var sub_tree = get_sub_tree_by_path(path);
        display_sub_tree(sub_tree, target, '');
    }

    function display_tree () {
        var file_list = $(file_list_jq),
        $speedTarget = $('<ul id="' + tree_root_id + '"></ul>');
        file_list.children().remove();
        display_tree_by_path(tree_delimeter, $speedTarget);
        file_list.append($speedTarget);
    }

    function load_files (f) {
        files = f;
        load_tree();
        display_tree();
    }

    function adjust_clip (e) {
        var offset = $(this).offset();
        if (zc_clip) {
            zc_clip.htmlBridge.style.top = offset.top + 'px';
            zc_clip.htmlBridge.style.left = offset.left + 'px';
        }
    }

    function create_clip () {
        zc_clip = new ZeroClipboard($(message_jq).find('.click-copy'), {moviePath:plugin_url + 'swf/ZeroClipboard.swf'});
    }

    function copy_paste_skeleton (target_text) {
        var copy_jq_el, copy_tmpl = '<span class="click-copy" title="Click to copy." data-clipboard-text="{{TEXT}}">Copy</span>';
        if (win.ZeroClipboard && plugin_url) {
            copy_jq_el = $(copy_tmpl.replace(/\{\{TEXT\}\}/g, target_text.replace(/'/g, '\'')));
            return copy_jq_el;
        }
        return null;
    }

    function display_message (message, error, jq_el) {
        var msg_wrap = $('<div>').html(message);
        if (jq_el) {
            msg_wrap.prepend(jq_el);
        }
        if (error) {
            msg_wrap.addClass(error_class);
        }
        $(message_jq).append(msg_wrap);
    }

    function clear_messages () {
        if (zc_clip) {
            ZeroClipboard.destroy();
            zc_clip = undefined;
        }
        $(message_jq).children().remove();
    }

    function load_container_picker () {
        var i = 0, sel = $('#sbts_cf_uploader_container'), size, mbgb = 'Mb',
        tmpl = '<option value="{{NAME}}">{{NAME}} ({{COUNT}}) [{{SIZE}}]</option>',
        cl = containers ? containers.length : 0;
        for (; i < cl; i += 1) {
            size = containers[i].bytes / (1024.00 * 1024.00);
            if (Math.floor(size / 1024) > 0) {
                size = size / 1024.00;
                mbgb = 'Gb';
            } else {
                mbgb = 'Mb';
            }
            containers[i].display_size = size.toFixed(2) + ' ' + mbgb;
            sel.append(tmpl.replace(/\{\{NAME\}\}/gi, containers[i].name).replace(/\{\{COUNT\}\}/, containers[i].count).replace(/\{\{SIZE\}\}/, containers[i].display_size));
        }
    }

    function load_containers (conts) {
        containers = conts;
        load_container_picker();
    }

    function get_files (container_to_get) {
        if (container_to_get) {
            $.ajax({
                'url'   : ajaxurl,
                'data'  : {
                    'action'        : 'get_files',
                    'sbts_cf_auth'  : 'add-later :)',
                    'sbts_cf_cont'  : container_to_get
                },
                'dataType'  : 'json',
                'beforeSend': function () {
                    $(file_list_jq).html('<p><img style="position:relative;top:4px;" src="/wp-admin/images/wpspin_light.gif" alt /> &nbsp;Loading file tree...</p>');
                },
                'success'   :   function (rjson) {
                    if (rjson && rjson.success) {
                        display_message(rjson.message, false);
                        load_files(rjson.pl);
                    } else {
                        display_message(rjson.message || "No files returned.", true);
                    }
                },
                'error' :   function (jqXHR, textStatus, errorThrown) {
                    display_message(errorThrown, true);
                    $(file_list_jq).html('');
                }
            });
        } else {
            display_message('No containers found.', true);
        }
    }

    function get_containers () {
        clear_messages();
        $.ajax({
            'url'   : ajaxurl,
            'data'  : {
                'action'        : 'get_containers',
                'sbts_cf_auth'  : 'add-later :)'
            },
            'dataType'  :   'json',
            'beforeSend': function () {
                $(file_list_jq).html('<p><img style="position:relative;top:4px;" src="/wp-admin/images/wpspin_light.gif" alt /> &nbsp;Loading containers...</p>');
            },
            'success'   :   function (rjson) {
                if (rjson && rjson.success) {
                    load_containers(rjson.pl);
                    display_message(rjson.message, false);
                    get_files($(cont_pick_jq).val());
                } else {
                    display_message(rjson ? rjson.message || 'No containers available.' : 'Could not connect.', true);
                    $(file_list_jq).html('');
                }
            },
            'error' :   function (jqXHR, textStatus, errorThrown) {
                display_message(errorThrown, true);
            }
        });
    }

    function upload_files (uploads) {
        var i = 0, uploads_len = uploads.length, u_type,
        cont = $(cont_pick_jq).val(),
        path = $(full_path_jq).data('full_path') || '',
        a_tmpl = '<a download class="upload-link" target="_blank" href="{{URI}}">{{NAME}}</a>',
        success_msg,
        form_data = tests.formdata ? new FormData() : null;
        if (tests.formdata && (path.length > 0)) {
            clear_messages();
            for (; i < uploads_len; i+= 1) {
                u_type = uploads[i].name.substr(uploads[i].name.lastIndexOf('.'));
                if (accepted_type[u_type]) {
                    form_data.append('file[]', uploads[i]);
                } else {
                    display_message('File type ' + u_type.replace('.', '') + ' not permitted.', true);
                }
            }
            form_data.append('action', 'upload_files');
            form_data.append('sbts_cf_auth', 'add-later :)');
            form_data.append('sbts_cf_cont', cont);
            form_data.append('sbts_cf_path', path);

            $(progress_bar_jq).css('width', '0%');
            $.ajax({
                'url'       :   ajaxurl,
                'type'      :   'post',
                'data'      :   form_data,
                'dataType'  :   'json',
                'xhr'       :   function () {
                    var myXHR = $.ajaxSettings.xhr();
                    if (myXHR.upload) {
                        myXHR.upload.addEventListener('progress', progress_handler, false);
                    }
                    return myXHR;
                },
                'beforeSend':   function () {
                    display_message('<img style="display:inline;padding-right:4px;margin:0;" src="/wp-admin/images/wpspin_light.gif" alt />Uploading ...', false);
                },
                'success'   :   function (rjson) {
                    var i = 0, up_len;
                    clear_messages();
                    $(progress_bar_jq).css('width', '0%');
                    if (rjson && rjson.success) {
                        display_message(rjson.message, false);
                        // TODO important - use load_files or the like to update files, tree and display
                        for (i = 0; i < rjson.pl.length; i += 1) {
                            success_msg = a_tmpl.replace(/\{\{URI\}\}/g, rjson.pl[i].uri).replace(/\{\{NAME\}\}/g, rjson.pl[i].name);
                            display_message(success_msg, false, copy_paste_skeleton(rjson.pl[i].uri));
                        }
                        create_clip();
                    } else if (rjson && rjson.success === false) {
                        // TODO should probably reload all files by calling get_files so
                        //      that if the user tried to upload many files and only a few
                        //      made it then they can check them
                        //      Same with the 'error' function
                        display_message(rjson.message, true);
                        display_message('Trying to reload files in case some files successfully uploaded.', true);
                        get_files(cont);
                    }
                },
                'error'     :   function (jqXHR, textStatus, errorThrown) {
                    $(progress_bar_jq).css('width', '0%');
                    display_message(errorThrown, true);
                },
                'cache': false,
                'contentType': false,
                'processData': false
            });
        } else if (path.length === 0) {
            display_message('Please add upload path.', true);
            $(path_display).addClass(error_class);
            error_no_path = true;
        }
    }

    function delete_files (form_data) {
        var cont = $(cont_pick_jq).val();
        if (tests.formdata && (form_data instanceof FormData)) {
            clear_messages();
            form_data.append('action', 'delete_files');
            form_data.append('sbts_cf_auth', 'add-later :)');
            form_data.append('sbts_cf_cont', cont);

            $.ajax({
                'url'       :   ajaxurl,
                'type'      :   'post',
                'data'      :   form_data,
                'dataType'  :   'json',
                'beforeSend':   function () {
                    display_message('<img style="display:inline;padding-right:4px;padding-top:4px;margin:0;" src="/wp-admin/images/wpspin_light.gif" alt />Deleting ...', false);
                },
                'success'   :   function (rjson) {
                    var i = 0, del_len;
                    clear_messages();
                    if (rjson && rjson.success) {
                        display_message(rjson.message, false);
                        if (rjson.pl.length) {
                            del_len = rjson.pl.length;
                            for(; i < del_len; i += 1) {
                                display_message(rjson.pl[i].name, false);
                                // TODO update files, tree and display
                                $(tree_root_jq).find('[data-file_name="' + rjson.pl[i].full_name + '"]').closest('.file-listing-wrap').remove();
                            }
                        }
                    } else if (rjson && (!rjson.success)) {
                        display_message(rjson.message, true);
                    }
                },
                'error'     :   function (jqXHR, textStatus, errorThrown) {
                    display_message(errorThrown, true);
                },
                'cache'     : false,
                'contentType':false,
                'processData':false
            });
        }
    }

    if ((tests.filereader === false) || (tests.formdata === false) || (tests.dnd === false)) {
        $('.fallbacks').css('display', 'block');
        $('.space-age').css('display', 'none');
        doc.getElementById('old_upload').onchange = function () {
            upload_files(this.files);
        };
    } else {
        holder.ondragover = function () { this.className = 'hover'; return false; };
        holder.ondragend = function () { this.className = ''; return false; };
        holder.ondrop = function (e) {
            this.className = '';
            e.preventDefault();
            upload_files(e.dataTransfer.files);
        };
    }

    function path_add_extra (existing, user_added) {
        if (user_added && user_added.substr(0, 1) === tree_delimeter) {
            user_added = user_added.substr(1);
        }
        if (user_added && user_added.substr(user_added.length - 1, 1) === tree_delimeter) {
            user_added = user_added.substr(0, user_added.length - 1);
        }
        user_added = encodeURI(user_added).replace(/%20/g, ' ');
        return (tree_delimeter === existing ? existing : existing + tree_delimeter) + (user_added ? user_added + tree_delimeter : '');
    }
    function update_upload_path (existing_path, user_added) {
        var full_path = path_add_extra(existing_path, user_added);
        $(full_path_jq).data('existing_path', existing_path);
        $(full_path_jq).data('full_path', full_path);
        $(full_path_jq).text(full_path);
        if (error_no_path && (full_path.length > 0)) {
            $(path_display_jq).removeClass(error_class);
            error_no_path = false;
        }
    }
    $(user_added_path_jq).on('keyup', function (e) {
        update_upload_path($(full_path_jq).data('existing_path'), $(this).val() || '');
    });
    $('.file-list').on('change', '.path-select', function (e) {
        update_upload_path($(this).data('existing_path'), $(user_added_path_jq).val() || '');
    });
    $('.file-list').on('click', '.path-choice>label', function (e) {
        $(this).closest('.path-choice').children('.sub-tree').toggle();
    });
    $('.file-list').on('click', '.delete-button', function (e) {
        var form_data, btn = $(this);
        if (tests.formdata && (btn.data('file_name').length > 0)) {
            if ('delete' === btn.text().toLowerCase()) {
                btn.text('Confirm?');
                btn.data('expire_delete', true);
                setTimeout(function expire_delete () {
                    if (btn.data('expire_delete')) {
                        btn.text('Delete');
                    }
                }, 2000);
            } else if ('confirm?' === btn.text().toLowerCase()) {
                btn.data('expire_delete', false);
                btn.text('Deleting...');
                form_data = new FormData();
                form_data.append('file[]', btn.data('file_name'));
                delete_files(form_data);
            }
        }
    });
    $(cont_pick_jq).on('change', function (e) {
        clear_messages();
        get_files($(cont_pick_jq).val());
    });
    $(message_jq).on('mouseenter', '.click-copy', adjust_clip);
    get_containers();
})(window, document, jQuery);
